{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "66a81706",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import pandas as pd\n",
    "import itertools\n",
    "import requests\n",
    "import zipfile\n",
    "import json\n",
    "import sys\n",
    "import io\n",
    "import re\n",
    "import os\n",
    "\n",
    "SUBSTRUCTURE_DATA = \"Substructures-DB-latest.txt\"\n",
    "CHEMONTID_DICTIONARY = \"CHEMONTID-mapper.json\"\n",
    "\n",
    "ONT_URL = 'http://classyfire.wishartlab.com/system/downloads/1_0/chemont/ChemOnt_2_1.obo.zip'\n",
    "ONT_FILE = 'ChemOnt_2_1.obo'\n",
    "ONT_DELIMITER = '[Term]'\n",
    "\n",
    "SDF_URL = 'https://hmdb.ca/system/downloads/current/structures.zip'\n",
    "SDF_FILE = 'structures.sdf'\n",
    "SDF_DELIMITER = '$$$$'\n",
    "ENCODING = 'utf-8'\n",
    "CLASSYFIRE_HMDB_URL = 'http://classyfire.wishartlab.com/system/downloads/1_0/datasets/HMDB_36_classyfire_21_annotations.csv.zip'\n",
    "CLASSYFIRE_HMDB_FILE = 'HMDB_36_classyfire_21_annotations.csv'\n",
    "CLASSYFIRE_CHEBI_URL = 'http://classyfire.wishartlab.com/system/downloads/1_0/datasets/ChEBI_126_classyfire_21_annotations.csv.zip'\n",
    "CLASSYFIRE_CHEBI_FILE = 'ChEBI_126_classyfire_21_annotations.csv'\n",
    "\n",
    "HMDB_KEY = 'HMDB_ID'\n",
    "ID_KEY = 'DATABASE_ID'\n",
    "DATABASE_KEY = 'DATABASE_NAME'\n",
    "SMILES_KEY = 'SMILES'\n",
    "FORMULA_KEY = 'FORMULA'\n",
    "IUPAC_KEY = 'JCHEM_IUPAC'\n",
    "NAME_KEY = 'GENERIC_NAME'\n",
    "SYNONYMS_KEY = 'SYNONYMS'\n",
    "SYNONYM_DELIMITER = ';'\n",
    "\n",
    "CLASSYFIER_ID_COLUMN = 0\n",
    "CLASSYFIER_ONT_COLUMN = 1\n",
    "CLASSYFIER_CLASS_COLUMN = 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e6e43c0",
   "metadata": {},
   "source": [
    "## Classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f0661e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Metabolite:\n",
    "    def __init__(\n",
    "            self,\n",
    "            record):\n",
    "        self.sdf_record = record\n",
    "        self.hmdb_id = self.lookup_sdf_info(HMDB_KEY)\n",
    "        self.iupac_id = self.lookup_sdf_info(IUPAC_KEY)\n",
    "        self.common_name = self.lookup_sdf_info(NAME_KEY)\n",
    "        self.smiles = self.lookup_sdf_info(SMILES_KEY)\n",
    "        self.synonyms = self.lookup_sdf_info(SYNONYMS_KEY)\n",
    "        self.taxonomy_ids = []\n",
    "        self.taxonomy_terms = []\n",
    "\n",
    "        del self.sdf_record\n",
    "            \n",
    "    def lookup_sdf_info(\n",
    "            self,\n",
    "            _key,\n",
    "            _delimiter=SYNONYM_DELIMITER):\n",
    "        \"\"\"\n",
    "        \"\"\"\n",
    "\n",
    "        for index in range(len(self.sdf_record)):\n",
    "            if _key in self.sdf_record[index]:\n",
    "                contents = self.sdf_record[index + 1] \\\n",
    "                    .replace('\\n', '') \\\n",
    "                    .replace(' ', '')\n",
    "                if _delimiter in contents:\n",
    "                    contents = contents.split(_delimiter)\n",
    "                return contents"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd215d6e",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cebf18c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_zip_archive(\n",
    "        output,\n",
    "        zip_url,\n",
    "        file_name):\n",
    "    \"\"\" Download HMDB SDF file containing all database SDF records\n",
    "    \"\"\"\n",
    "    output_file = os.path.join(output, file_name)\n",
    "    this_url = requests.get(zip_url)\n",
    "    if this_url.ok:\n",
    "        print(\"Unzipping HMDB metabolite SDF reference...\")\n",
    "        this_zip = zipfile.ZipFile(io.BytesIO(this_url.content))\n",
    "        this_zip.extractall(output)\n",
    "        this_zip = None\n",
    "    else:\n",
    "        raise Exception(\"Unable to download file at: \" + zip_url)\n",
    "    return output_file\n",
    "\n",
    "\n",
    "def parse_records_as_list(\n",
    "        sdf_file,\n",
    "        _delimiter=SDF_DELIMITER,\n",
    "        _encoding=ENCODING):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    output_records = []\n",
    "    with open(sdf_file, \"r\", encoding=_encoding) as _f:\n",
    "        for key, group in itertools.groupby(_f, lambda _l: _delimiter in _l):\n",
    "            if not key:\n",
    "                group = [x for x in list(group) if x != '\\n']\n",
    "                output_records.append(group)\n",
    "    return output_records\n",
    "\n",
    "\n",
    "def parse_records_as_datatable(\n",
    "        input_file,\n",
    "        _delimiter=','):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    output_records = pd.read_csv(\n",
    "        input_file,\n",
    "        sep=_delimiter,\n",
    "        error_bad_lines=False,\n",
    "        warn_bad_lines=True)\n",
    "    return output_records\n",
    "\n",
    "\n",
    "def dict_from_datatable(\n",
    "        input_source,\n",
    "        id_col=CLASSYFIER_ID_COLUMN,\n",
    "        ont_col=CLASSYFIER_ONT_COLUMN):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    table_dict = {}\n",
    "    for index, row in input_source.iterrows():\n",
    "        if row[id_col] in table_dict:\n",
    "            table_dict[row[id_col]].append(row[CLASSYFIER_ONT_COLUMN])\n",
    "        else:\n",
    "            table_dict[row[id_col]] = [row[CLASSYFIER_ONT_COLUMN]]\n",
    "    return table_dict\n",
    "\n",
    "\n",
    "def make_ontology_dictionary(\n",
    "        hmdb_source,\n",
    "        chebi_source,\n",
    "        ont_col=CLASSYFIER_ONT_COLUMN,\n",
    "        name_col=CLASSYFIER_CLASS_COLUMN):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "\n",
    "    def add_ontology(\n",
    "            ont_dict,\n",
    "            data_source,\n",
    "            ont_col,\n",
    "            name_col):\n",
    "        for index, row in data_source.iterrows():\n",
    "            if row[ont_col] in ontology_dictionary:\n",
    "                if ont_dict[row[ont_col]] != row[name_col]:\n",
    "                    print(\n",
    "                        'Mismatched ID & Name for: '\n",
    "                        + str(row[ont_col])\n",
    "                        + '\\n- ' + str(ont_dict[row[ont_col]])\n",
    "                        + '\\n- ' + str(row[name_col])\n",
    "                        + '\\n')\n",
    "            else:\n",
    "                ont_dict[row[ont_col]] = row[name_col]\n",
    "        return ont_dict\n",
    "\n",
    "    ontology_dictionary = {}\n",
    "    ontology_dictionary = add_ontology(\n",
    "        ontology_dictionary,\n",
    "        chebi_source,\n",
    "        ont_col,\n",
    "        name_col)\n",
    "    ontology_dictionary = add_ontology(\n",
    "        ontology_dictionary,\n",
    "        hmdb_source,\n",
    "        ont_col,\n",
    "        name_col)\n",
    "    print('Found ' + str(len(ontology_dictionary.keys())) + ' terms')\n",
    "    return ontology_dictionary\n",
    "\n",
    "\n",
    "def parse_metadata_from_records(\n",
    "        records):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    output_metadata = []\n",
    "    for r in records:\n",
    "        thisSDF = Metabolite(record=r)\n",
    "        output_metadata.append(thisSDF)\n",
    "    return output_metadata\n",
    "\n",
    "\n",
    "def retrieve_substructures(\n",
    "        records,\n",
    "        hmdb_dictionary,\n",
    "        chebi_dictionary,\n",
    "        ontology_dictionary):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    for r in range(len(records)):\n",
    "        base_id = records[r].hmdb_id \\\n",
    "            .replace('HMDB', '') \\\n",
    "            .lstrip('0')\n",
    "        for x in range(12):\n",
    "            this_id = ('HMDB' + base_id.zfill(x + 1))\n",
    "            if this_id in hmdb_dictionary:\n",
    "                records[r].taxonomy_ids = hmdb_dictionary[this_id]\n",
    "                for t in records[r].taxonomy_ids:\n",
    "                    if t in ontology_dictionary:\n",
    "                        records[r].taxonomy_terms \\\n",
    "                            .append(ontology_dictionary[t])\n",
    "                break\n",
    "    return records\n",
    "\n",
    "\n",
    "def write_output(\n",
    "        output_database,\n",
    "        output_location,\n",
    "        synonym_key='synonyms',\n",
    "        tax_id_key='taxonomy_ids',\n",
    "        tax_term_key='taxonomy_terms'):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    def try_join(\n",
    "            l,\n",
    "            sep=';'):\n",
    "        try:\n",
    "            return str(sep).join(map(str, l))\n",
    "        except TypeError:\n",
    "            return ''\n",
    "\n",
    "    mod_output = [d.__dict__ for d in output_database]\n",
    "    output_table = pd.DataFrame(mod_output)\n",
    "    output_table[synonym_key] = [\n",
    "        try_join(l) for l in output_table[synonym_key]]\n",
    "    output_table[tax_id_key] = [\n",
    "        try_join(l) for l in output_table[tax_id_key]]\n",
    "    output_table[tax_term_key] = [\n",
    "        try_join(l) for l in output_table[tax_term_key]]\n",
    "    output_table.to_csv(\n",
    "        os.path.join(\n",
    "            str(output_location), SUBSTRUCTURE_DATA),\n",
    "        sep='\\t')\n",
    "\n",
    "\n",
    "def write_dictionary(\n",
    "        output_dictionary,\n",
    "        output_location,\n",
    "        output_name=CHEMONTID_DICTIONARY):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    with open(os.path.join(output_location, output_name), 'w') as fp:\n",
    "        json.dump(output_dictionary, fp)\n",
    "\n",
    "\n",
    "def clean_downloads(\n",
    "        output_files):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    for o in output_files:\n",
    "        try:\n",
    "            os.remove(o)\n",
    "        except:\n",
    "            print('Unable to remove file ' + o)\n",
    "\n",
    "\n",
    "def import_table(\n",
    "        _path,\n",
    "        _file,\n",
    "        index_col=None):\n",
    "    \"\"\"Import URL as tab-delimited table\n",
    "    \"\"\"\n",
    "\n",
    "    return pd.read_csv(\n",
    "        os.path.join(_path, _file),\n",
    "        sep='\\t',\n",
    "        index_col=index_col,\n",
    "        low_memory=False\n",
    "    )\n",
    "\n",
    "\n",
    "def import_json(\n",
    "        _path,\n",
    "        _file):\n",
    "    \"\"\"Import URL as JSON-formatted dictionary\n",
    "    \"\"\"\n",
    "    with open(os.path.join(_path, _file)) as json_file:\n",
    "        data = json.load(json_file)\n",
    "\n",
    "    return data\n",
    "\n",
    "\n",
    "def parse_ont_as_list(\n",
    "        ont_file,\n",
    "        _delimiter=ONT_DELIMITER,\n",
    "        _encoding=ENCODING):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    output_records = []\n",
    "    with open(ont_file, \"r\", encoding=_encoding) as _f:\n",
    "        for key, group in itertools.groupby(_f, lambda _l: _delimiter in _l):\n",
    "            if not key:\n",
    "                group = [x for x in list(group) if x != '\\n']\n",
    "                output_records.append(group)\n",
    "    return output_records\n",
    "\n",
    "def parse_dict_from_records(\n",
    "        ont_data):\n",
    "\n",
    "    ont_dict = {}\n",
    "    for x in ont_data:\n",
    "        id = \"\"\n",
    "        name = \"\"\n",
    "        is_a = \"\"\n",
    "        \n",
    "        _x = [i for i in x if \"is_a\"]\n",
    "        if len(_x) > 0:\n",
    "            for i in x:\n",
    "                if \"id: \" in i:\n",
    "                    id = i.split(\"id: \")[1].split(\"\\n\")[0]\n",
    "                if \"name: \" in i:\n",
    "                    name = i.split(\"name: \")[1].split(\"\\n\")[0]\n",
    "                if \"is_a: \" in i:\n",
    "                    is_a = i.split(\"is_a: \")[1].split(\" ! \")[0]\n",
    "        \n",
    "        ont_dict[id] = {\n",
    "            \"id\": id,\n",
    "            \"name\": name,\n",
    "            \"is_a\": is_a\n",
    "        }\n",
    "    \n",
    "    return ont_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fcb2e011",
   "metadata": {},
   "outputs": [],
   "source": [
    "args_dict = {\n",
    "    \"output\": os.getcwd()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "360d8d4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unzipping HMDB metabolite SDF reference...\n"
     ]
    }
   ],
   "source": [
    "# Parse all HMDB SDF records\n",
    "sdf_file = download_zip_archive(\n",
    "    output=args_dict[\"output\"],\n",
    "    zip_url=SDF_URL,\n",
    "    file_name=SDF_FILE)\n",
    "record_list = parse_records_as_list(\n",
    "    sdf_file=sdf_file)\n",
    "output_metadata = parse_metadata_from_records(\n",
    "    records=record_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "47d42ee2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unzipping HMDB metabolite SDF reference...\n"
     ]
    }
   ],
   "source": [
    "# Parse ClassyFire HMDB records\n",
    "cf_hmdb_file = download_zip_archive(\n",
    "    output=args_dict[\"output\"],\n",
    "    zip_url=CLASSYFIRE_HMDB_URL,\n",
    "    file_name=CLASSYFIRE_HMDB_FILE)\n",
    "hmdb_records = parse_records_as_datatable(\n",
    "    input_file=cf_hmdb_file)\n",
    "hmdb_dictionary = dict_from_datatable(\n",
    "    input_source=hmdb_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b781de78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CHEMONTID:0003224',\n",
       " 'CHEMONTID:0002960',\n",
       " 'CHEMONTID:0002987',\n",
       " 'CHEMONTID:0001090',\n",
       " 'CHEMONTID:0000129',\n",
       " 'CHEMONTID:0003457',\n",
       " 'CHEMONTID:0002449',\n",
       " 'CHEMONTID:0000347',\n",
       " 'CHEMONTID:0000013',\n",
       " 'CHEMONTID:0001262',\n",
       " 'CHEMONTID:0004139',\n",
       " 'CHEMONTID:0000436',\n",
       " 'CHEMONTID:0001878',\n",
       " 'CHEMONTID:0000011',\n",
       " 'CHEMONTID:0001831',\n",
       " 'CHEMONTID:0003872',\n",
       " 'CHEMONTID:0000475',\n",
       " 'CHEMONTID:0001093',\n",
       " 'CHEMONTID:0000265',\n",
       " 'CHEMONTID:9999999',\n",
       " 'CHEMONTID:0001143',\n",
       " 'CHEMONTID:0001346',\n",
       " 'CHEMONTID:0003909',\n",
       " 'CHEMONTID:0003327',\n",
       " 'CHEMONTID:0000331',\n",
       " 'CHEMONTID:0002105',\n",
       " 'CHEMONTID:0002203',\n",
       " 'CHEMONTID:0004144',\n",
       " 'CHEMONTID:0004150',\n",
       " 'CHEMONTID:0000078',\n",
       " 'CHEMONTID:0001797',\n",
       " 'CHEMONTID:0003886',\n",
       " 'CHEMONTID:0000012',\n",
       " 'CHEMONTID:0003458',\n",
       " 'CHEMONTID:0001541',\n",
       " 'CHEMONTID:0001540',\n",
       " 'CHEMONTID:0001096',\n",
       " 'CHEMONTID:0002311',\n",
       " 'CHEMONTID:0000289',\n",
       " 'CHEMONTID:0000264',\n",
       " 'CHEMONTID:0000000',\n",
       " 'CHEMONTID:0004707',\n",
       " 'CHEMONTID:0003940',\n",
       " 'CHEMONTID:0000463',\n",
       " 'CHEMONTID:0004603',\n",
       " 'CHEMONTID:0000402',\n",
       " 'CHEMONTID:0001804',\n",
       " 'CHEMONTID:0000002',\n",
       " 'CHEMONTID:0000278',\n",
       " 'CHEMONTID:0000323',\n",
       " 'CHEMONTID:0004557',\n",
       " 'CHEMONTID:0000004',\n",
       " 'CHEMONTID:0004140',\n",
       " 'CHEMONTID:0001967',\n",
       " 'CHEMONTID:0002259',\n",
       " 'CHEMONTID:0001497',\n",
       " 'CHEMONTID:0000408',\n",
       " 'CHEMONTID:0002450',\n",
       " 'CHEMONTID:0002453',\n",
       " 'CHEMONTID:0001506',\n",
       " 'CHEMONTID:0002209',\n",
       " 'CHEMONTID:0000026',\n",
       " 'CHEMONTID:0001618',\n",
       " 'CHEMONTID:0001544',\n",
       " 'CHEMONTID:0000245',\n",
       " 'CHEMONTID:0000075',\n",
       " 'CHEMONTID:0004404',\n",
       " 'CHEMONTID:0001661',\n",
       " 'CHEMONTID:0001663',\n",
       " 'CHEMONTID:0002310',\n",
       " 'CHEMONTID:0003260',\n",
       " 'CHEMONTID:0001112',\n",
       " 'CHEMONTID:0001200']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hmdb_dictionary[\"HMDB01166\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "951700d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unzipping HMDB metabolite SDF reference...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "b'Skipping line 775780: expected 4 fields, saw 6\\n'\n",
      "b'Skipping line 1705535: expected 4 fields, saw 5\\n'\n"
     ]
    }
   ],
   "source": [
    "# Parse ClassyFire CHEBI records\n",
    "cf_chebi_file = download_zip_archive(\n",
    "    output=args_dict[\"output\"],\n",
    "    zip_url=CLASSYFIRE_CHEBI_URL,\n",
    "    file_name=CLASSYFIRE_CHEBI_FILE)\n",
    "chebi_records = parse_records_as_datatable(\n",
    "    input_file=cf_chebi_file)\n",
    "chebi_dictionary = dict_from_datatable(\n",
    "    input_source=chebi_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8ad24573",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Mismatched ID & Name for: nan\n",
      "- nan\n",
      "- nan\n",
      "\n",
      "Found 3250 terms\n"
     ]
    }
   ],
   "source": [
    "# Make ontology reference\n",
    "ontology_dictionary = make_ontology_dictionary(\n",
    "    hmdb_source=hmdb_records,\n",
    "    chebi_source=chebi_records)\n",
    "\n",
    "write_dictionary(\n",
    "    output_dictionary=ontology_dictionary,\n",
    "    output_location=args_dict[\"output\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7ec0860d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add ontology IDs and terms to each SDF record\n",
    "output_substructures = retrieve_substructures(\n",
    "    records=output_metadata,\n",
    "    hmdb_dictionary=hmdb_dictionary,\n",
    "    chebi_dictionary=chebi_dictionary,\n",
    "    ontology_dictionary=ontology_dictionary)\n",
    "\n",
    "# Output and clean\n",
    "write_output(\n",
    "    output_database=output_substructures,\n",
    "    output_location=args_dict[\"output\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8b89eb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "45137ef5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3-Hydroxybutyryl-CoA\n",
      "HMDB0001166\n",
      "['(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters', '(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters', '(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters', '(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters']\n"
     ]
    }
   ],
   "source": [
    "for r in range(len(output_metadata)):\n",
    "    base_id = output_metadata[r].hmdb_id \\\n",
    "        .replace('HMDB', '') \\\n",
    "        .lstrip('0')\n",
    "    if base_id == \"1166\":\n",
    "        print(output_metadata[r].common_name)\n",
    "        print(output_metadata[r].hmdb_id)\n",
    "    for x in range(12):\n",
    "        this_id = ('HMDB' + base_id.zfill(x + 1))\n",
    "        if this_id in hmdb_dictionary:\n",
    "            output_metadata[r].taxonomy_ids = hmdb_dictionary[this_id]\n",
    "            for t in output_metadata[r].taxonomy_ids:\n",
    "                if t in ontology_dictionary:\n",
    "                    output_metadata[r].taxonomy_terms \\\n",
    "                        .append(ontology_dictionary[t])\n",
    "            \n",
    "            break\n",
    "    \n",
    "    if base_id == \"1166\":\n",
    "        print(output_metadata[r].taxonomy_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c87dbee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7309c754",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f93216c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "903cc547",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'hmdb_id': 'HMDB0001166', 'iupac_id': '{[(2R,3S,4R,5R)-5-(6-amino-9H-purin-9-yl)-4-hydroxy-2-({[hydroxy({hydroxy[3-hydroxy-3-({2-[(2-{[(3R)-3-hydroxybutanoyl]sulfanyl}ethyl)carbamoyl]ethyl}carbamoyl)-2,2-dimethylpropoxy]phosphoryl}oxy)phosphoryl]oxy}methyl)oxolan-3-yl]oxy}phosphonicacid', 'common_name': '3-Hydroxybutyryl-CoA', 'smiles': 'C[C@@H](O)CC(=O)SCCNC(=O)CCNC(=O)C(O)C(C)(C)COP(O)(=O)OP(O)(=O)OC[C@H]1O[C@H]([C@H](O)[C@@H]1OP(O)(O)=O)N1C=NC2=C1N=CN=C2N', 'synonyms': ['(3R)-3-Hydroxybutanoyl-CoA', '(3R)-3-Hydroxybutanoyl-CoenzymeA', '(R)-3-Hydroxybutanoyl-CoA', '(R)-3-Hydroxybutanoyl-CoenzymeA', '(S)-3-hydroxybutanoyl-CoA', '(S)-3-hydroxybutanoyl-CoenzymeA', '3-Hydroxybutanoyl-CoA', '3-Hydroxybutanoyl-coenzymea', '3-Hydroxybutyryl-CoA', '3-Hydroxybutyryl-coenzymeA', '3-OH-butyryl-CoA', '3-OH-butyryl-CoenzymeA', 'beta-Hydroxybutyryl-CoA', 'beta-Hydroxybutyryl-CoenzymeA', 'beta-Hydroxybutyryl-S-CoA', 'beta-Hydroxybutyryl-S-CoenzymeA', 'Hydroxy-butyryl-CoA', 'Hydroxy-butyryl-CoenzymeA'], 'taxonomy_ids': ['CHEMONTID:0003224', 'CHEMONTID:0002960', 'CHEMONTID:0002987', 'CHEMONTID:0001090', 'CHEMONTID:0000129', 'CHEMONTID:0003457', 'CHEMONTID:0002449', 'CHEMONTID:0000347', 'CHEMONTID:0000013', 'CHEMONTID:0001262', 'CHEMONTID:0004139', 'CHEMONTID:0000436', 'CHEMONTID:0001878', 'CHEMONTID:0000011', 'CHEMONTID:0001831', 'CHEMONTID:0003872', 'CHEMONTID:0000475', 'CHEMONTID:0001093', 'CHEMONTID:0000265', 'CHEMONTID:9999999', 'CHEMONTID:0001143', 'CHEMONTID:0001346', 'CHEMONTID:0003909', 'CHEMONTID:0003327', 'CHEMONTID:0000331', 'CHEMONTID:0002105', 'CHEMONTID:0002203', 'CHEMONTID:0004144', 'CHEMONTID:0004150', 'CHEMONTID:0000078', 'CHEMONTID:0001797', 'CHEMONTID:0003886', 'CHEMONTID:0000012', 'CHEMONTID:0003458', 'CHEMONTID:0001541', 'CHEMONTID:0001540', 'CHEMONTID:0001096', 'CHEMONTID:0002311', 'CHEMONTID:0000289', 'CHEMONTID:0000264', 'CHEMONTID:0000000', 'CHEMONTID:0004707', 'CHEMONTID:0003940', 'CHEMONTID:0000463', 'CHEMONTID:0004603', 'CHEMONTID:0000402', 'CHEMONTID:0001804', 'CHEMONTID:0000002', 'CHEMONTID:0000278', 'CHEMONTID:0000323', 'CHEMONTID:0004557', 'CHEMONTID:0000004', 'CHEMONTID:0004140', 'CHEMONTID:0001967', 'CHEMONTID:0002259', 'CHEMONTID:0001497', 'CHEMONTID:0000408', 'CHEMONTID:0002450', 'CHEMONTID:0002453', 'CHEMONTID:0001506', 'CHEMONTID:0002209', 'CHEMONTID:0000026', 'CHEMONTID:0001618', 'CHEMONTID:0001544', 'CHEMONTID:0000245', 'CHEMONTID:0000075', 'CHEMONTID:0004404', 'CHEMONTID:0001661', 'CHEMONTID:0001663', 'CHEMONTID:0002310', 'CHEMONTID:0003260', 'CHEMONTID:0001112', 'CHEMONTID:0001200'], 'taxonomy_terms': ['(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters', '(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters', '(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters', '(R)-3-hydroxyacyl CoAs', '3-hydroxyacyl CoAs', '6-aminopurines', 'Acyl CoAs', 'Alcohols and polyols', 'Alkyl phosphates', 'Amines', 'Amino acids and derivatives', 'Amino acids, peptides, and analogues', 'Aminopyrimidines and derivatives', 'Azacyclic compounds', 'Azoles', 'Beta amino acids and derivatives', 'Carbohydrates and carbohydrate conjugates', 'Carbonyl compounds', 'Carbothioic S-esters', 'Carboxylic acid amides', 'Carboxylic acid derivatives', 'Carboxylic acids and derivatives', 'Chemical entities', 'Coenzyme A and derivatives', 'Diazines', 'Fatty Acyls', 'Fatty acyl thioesters', 'Fatty amides', 'Glycosyl compounds', 'Glycosylamines', 'Heteroaromatic compounds', 'Hydrocarbon derivatives', 'Imidazoles', 'Imidazopyrimidines', 'Imidolactams', 'Lipids and lipid-like molecules', 'Monoalkyl phosphates', 'Monosaccharide phosphates', 'Monosaccharides', 'N-acyl amines', 'N-substituted imidazoles', 'Nucleosides, nucleotides, and analogues', 'Organic acids and derivatives', 'Organic compounds', 'Organic nitrogen compounds', 'Organic oxides', 'Organic oxoanionic compounds', 'Organic oxygen compounds', 'Organic phosphoric acids and derivatives', 'Organic pyrophosphates', 'Organoheterocyclic compounds', 'Organonitrogen compounds', 'Organooxygen compounds', 'Organopnictogen compounds', 'Organosulfur compounds', 'Oxacyclic compounds', 'Oxolanes', 'Pentose phosphates', 'Pentoses', 'Phosphate esters', 'Primary amines', 'Primary aromatic amines', 'Purine nucleotides', \"Purine ribonucleoside 3',5'-bisphosphates\", 'Purine ribonucleoside bisphosphates', 'Purine ribonucleoside diphosphates', 'Purine ribonucleotides', 'Purines and purine derivatives', 'Pyrimidines and pyrimidine derivatives', \"Ribonucleoside 3'-phosphates\", 'Secondary alcohols', 'Secondary carboxylic acid amides', 'Substituted imidazoles', 'Sulfenyl compounds', 'Thiocarboxylic acids and derivatives', 'Thioesters']}\n"
     ]
    }
   ],
   "source": [
    "for x in output_substructures:\n",
    "    if x.hmdb_id == \"HMDB0001166\":\n",
    "        print(vars(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b9e540d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import parsed ClassyFire sub-structure database\n",
    "substructure_table = import_table(\n",
    "    _path=args_dict[\"output\"],\n",
    "    _file=SUBSTRUCTURE_DATA,\n",
    "    index_col=0)\n",
    "substructure_table.index = substructure_table['hmdb_id']\n",
    "substructure_dictionary = substructure_table.T.to_dict()\n",
    "\n",
    "write_dictionary(\n",
    "    output_dictionary=substructure_dictionary,\n",
    "    output_location=args_dict[\"output\"], \n",
    "    output_name=\"CHEMONTID-substructure-dictionary.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6d15d31f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unzipping HMDB metabolite SDF reference...\n"
     ]
    }
   ],
   "source": [
    "# Generate CHEMONTID hierarchal structure reference \n",
    "ont_file = download_zip_archive(\n",
    "    output=args_dict[\"output\"],\n",
    "    zip_url=ONT_URL,\n",
    "    file_name=ONT_FILE)\n",
    "ont_data = parse_ont_as_list(\n",
    "    ont_file=ont_file)\n",
    "ont_dict = parse_dict_from_records(\n",
    "    ont_data=ont_data)\n",
    "write_dictionary(\n",
    "    output_dictionary=ont_dict,\n",
    "    output_location=args_dict[\"output\"],\n",
    "    output_name='CHEMONTID-hierarchy-dictionary.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03d9b5fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_downloads(\n",
    "    output_files=[\n",
    "        sdf_file,\n",
    "        cf_hmdb_file,\n",
    "        cf_chebi_file,\n",
    "        ont_file])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
